File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int answer = -1;
    
    public int solution(int k, int[][] dungeons) {

        boolean[] visited = new boolean[dungeons.length];
        int cnt = 0;
        
        dfs(k, dungeons, visited, cnt);
        
        return answer;
    }
    
    public void dfs(int k, int[][] dungeons, boolean[] visited, int cnt){
        for(int i = 0; i < dungeons.length; i++){
            if(!visited[i] && k >= dungeons[i][0]){
                visited[i] = true;
                dfs(k-dungeons[i][1], dungeons, visited, cnt+1);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, cnt);
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    
    def play_dungeons(turn, now_k):
        nonlocal dungeons
        
        count = 0
        for t in turn:
            if now_k >= dungeons[t][0]:
                now_k -= dungeons[t][1]
                count +=1
        
        return count 
                
    turn = [i for i in range(len(dungeons))]
    for t in permutations(turn):
        answer = max(answer, play_dungeons(t, k))
    
    return answer


# 유저가 탐험할 수 있는 최대 던전을 구하라
# k: 현재 피로도
# 탐험할 수 있는가. 깊게 파고들어야 한다. dfs로 풀어보자?
# 1-2-3 또는 2-3-1 순서로해야할수도있다.